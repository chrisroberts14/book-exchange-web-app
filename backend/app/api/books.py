"""Module for book endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from starlette.status import HTTP_201_CREATED, HTTP_404_NOT_FOUND, HTTP_204_NO_CONTENT

from backend.app.api_models import BookOut, BookIn, BookPatch
from backend.app.core.db import get_db
from backend.app.db_models import BookDb

books = APIRouter()


@books.post("/", status_code=HTTP_201_CREATED)
async def create_book(book: BookIn, db: Session = Depends(get_db)) -> BookOut:
    """
    Create a book.

    :param db:
    :param book: book to create
    :return:
    """
    db_book = BookDb(**book.dict())
    return BookDb.create(db, db_book)


@books.get("/")
async def get_all_books(db: Session = Depends(get_db)) -> list[BookOut]:
    """
    Get all books.

    :return:
    """
    return BookDb.get_all(db)


@books.get("/{book_id}")
async def get_book(book_id: UUID, db: Session = Depends(get_db)) -> BookOut:
    """
    Get a book by id.

    :param book_id: id of the book
    :param db: database session
    :return:
    """
    book = BookDb.get_by_id(db, book_id)
    if book is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Book not found")
    return book


@books.patch("/{book_id}")
async def update_book(
    book_id: UUID, book_patch: BookPatch, db: Session = Depends(get_db)
) -> BookOut:
    """
    Update a book.

    :param book_id:
    :param book_patch:
    :param db:
    :return:
    """
    book = BookDb.get_by_id(db, book_id)
    if book is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Book not found")
    return BookDb.update(db, book_patch, book_id)


@books.delete("/{book_id}", status_code=HTTP_204_NO_CONTENT)
async def delete_book(book_id: UUID, db: Session = Depends(get_db)) -> None:
    """
    Delete a book.

    :param book_id:
    :param db:
    :return:
    """
    book = BookDb.get_by_id(db, book_id)
    if book is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Book not found")
    BookDb.delete(db, book_id)
