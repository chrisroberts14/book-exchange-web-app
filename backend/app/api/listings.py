"""Module for listings endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from starlette.status import HTTP_201_CREATED, HTTP_404_NOT_FOUND

from backend.app.api_models import ListingIn, ListingOut, ListingPatch
from backend.app.core.db import get_db
from backend.app.db_models import ListingDb

listings = APIRouter()


@listings.get("/", response_model=list[ListingOut])
async def get_listings(db: Session = Depends(get_db)) -> list[ListingOut]:
    """
    Get all listings.

    :return:
    """
    return ListingDb.get_all(db)


@listings.post("/", status_code=HTTP_201_CREATED)
async def create_listing(data: ListingIn, db: Session = Depends(get_db)) -> ListingOut:
    """
    Create a listing.

    :return:
    """
    return ListingDb.create(db, ListingDb(**data.dict()))


@listings.get("/{listing_id}", response_model=ListingOut)
async def get_listing(listing_id: UUID, db: Session = Depends(get_db)) -> ListingOut:
    """
    Get a listing by id.

    :param listing_id: id of the listing
    :param db: database session
    :return:
    """
    listing = ListingDb.get_by_id(db, listing_id)
    if listing is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Listing not found")
    return listing


@listings.patch("/{listing_id}")
async def update_listing(
    listing_id: UUID, data: ListingPatch, db: Session = Depends(get_db)
) -> ListingOut:
    """
    Update a listing.

    :return:
    """
    listing = ListingDb.get_by_id(db, listing_id)
    if listing is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Listing not found")
    return ListingDb.update(db, data, listing_id)
