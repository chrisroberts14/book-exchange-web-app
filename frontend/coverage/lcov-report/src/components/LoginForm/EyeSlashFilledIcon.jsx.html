
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/components/LoginForm/EyeSlashFilledIcon.jsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>

<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> / <a href="index.html">src/components/LoginForm</a> EyeSlashFilledIcon.jsx</h1>
        <div class='clearfix'>

            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>1/2</span>
            </div>


            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/0</span>
            </div>


            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>0/1</span>
            </div>


            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>1/2</span>
            </div>


        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React from "react";
&nbsp;
export const EyeSlashFilledIcon = <span class="fstat-no" title="function not covered" >(p</span>rops) =&gt; (
<span class="cstat-no" title="statement not covered" >    &lt;svg</span>
        aria-hidden="true"
        fill="none"
        focusable="false"
        height="1em"
        role="presentation"
        viewBox="0 0 24 24"
        width="1em"
        {...props}
    &gt;
        &lt;path
            d="M21.2714 9.17834C20.9814 8.71834 20.6714 8.28834 20.3514 7.88834C19.9814 7.41834 19.2814 7.37834 18.8614 7.79834L15.8614 10.7983C16.0814 11.4583 16.1214 12.2183 15.9214 13.0083C15.5714 14.4183 14.4314 15.5583 13.0214 15.9083C12.2314 16.1083 11.4714 16.0683 10.8114 15.8483C10.8114 15.8483 9.38141 17.2783 8.35141 18.3083C7.85141 18.8083 8.01141 19.6883 8.68141 19.9483C9.75141 20.3583 10.8614 20.5683 12.0014 20.5683C13.7814 20.5683 15.5114 20.0483 17.0914 19.0783C18.7014 18.0783 20.1514 16.6083 21.3214 14.7383C22.2714 13.2283 22.2214 10.6883 21.2714 9.17834Z"
            fill="currentColor"
        /&gt;
        &lt;path
            d="M14.0206 9.98062L9.98062 14.0206C9.47062 13.5006 9.14062 12.7806 9.14062 12.0006C9.14062 10.4306 10.4206 9.14062 12.0006 9.14062C12.7806 9.14062 13.5006 9.47062 14.0206 9.98062Z"
            fill="currentColor"
        /&gt;
        &lt;path
            d="M18.25 5.74969L14.86 9.13969C14.13 8.39969 13.12 7.95969 12 7.95969C9.76 7.95969 7.96 9.76969 7.96 11.9997C7.96 13.1197 8.41 14.1297 9.14 14.8597L5.76 18.2497H5.75C4.64 17.3497 3.62 16.1997 2.75 14.8397C1.75 13.2697 1.75 10.7197 2.75 9.14969C3.91 7.32969 5.33 5.89969 6.91 4.91969C8.49 3.95969 10.22 3.42969 12 3.42969C14.23 3.42969 16.39 4.24969 18.25 5.74969Z"
            fill="currentColor"
        /&gt;
        &lt;path
            d="M14.8581 11.9981C14.8581 13.5681 13.5781 14.8581 11.9981 14.8581C11.9381 14.8581 11.8881 14.8581 11.8281 14.8381L14.8381 11.8281C14.8581 11.8881 14.8581 11.9381 14.8581 11.9981Z"
            fill="currentColor"
        /&gt;
        &lt;path
            d="M21.7689 2.22891C21.4689 1.92891 20.9789 1.92891 20.6789 2.22891L2.22891 20.6889C1.92891 20.9889 1.92891 21.4789 2.22891 21.7789C2.37891 21.9189 2.56891 21.9989 2.76891 21.9989C2.96891 21.9989 3.15891 21.9189 3.30891 21.7689L21.7689 3.30891C22.0789 3.00891 22.0789 2.52891 21.7689 2.22891Z"
            fill="currentColor"
        /&gt;
    &lt;/svg&gt;
);
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2024-07-11T14:20:22.840Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
