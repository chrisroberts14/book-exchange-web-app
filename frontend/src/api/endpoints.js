const address = "http://localhost:8000";

export const loginEndpoint = address + "/auth/login";
export const signupEndpoint = address + "/users/";
export const getAllBooksEndpoint = address + "/books/";
export const getAllListingsEndpoint = address + "/listings/";
