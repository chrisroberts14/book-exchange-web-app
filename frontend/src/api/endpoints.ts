const address = "http://localhost:8000";

export const loginEndpoint: string = address + "/auth/login";
