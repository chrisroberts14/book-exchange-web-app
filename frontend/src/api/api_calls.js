import axios from 'axios';
import {loginEndpoint} from "./endpoints.js";


export const login = async (credentials) => {
    const response = await axios.post(loginEndpoint, new URLSearchParams({
        'username': credentials.username,
        'password': credentials.password
    }));
    return response.data;
}
