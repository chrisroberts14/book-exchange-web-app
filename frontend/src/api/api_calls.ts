import axios from 'axios';
import {loginEndpoint} from "./endpoints";

type LoginCredentials = {
    username: string,
    password: string
}

export const login = async (credentials: LoginCredentials) => {
    const response = await axios.post(loginEndpoint, new URLSearchParams({
        'username': credentials.username,
        'password': credentials.password
    }));
    console.log('Login response:', response.data);
    return response.data;
}
