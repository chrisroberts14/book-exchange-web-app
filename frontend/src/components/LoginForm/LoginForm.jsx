import React, {useState} from "react";
import { login } from "../../api/api_calls";
import { Input } from "@nextui-org/react";
import {EyeFilledIcon} from "../EyeFilledIcon/EyeFilledIcon.jsx";
import {EyeSlashFilledIcon} from "../EyeSlashFilledIcon/EyeSlashFilledIcon.jsx";
import {Button} from "@nextui-org/react";


export const LoginForm = ({user, setUser}) => {
    const [isVisible, setIsVisible] = React.useState(false);
    const toggleVisibility = () => setIsVisible(!isVisible);
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');


    const handleLogin = async (event) => {
        event.preventDefault();
        console.log('Logging in with:', username, password);
        // Even though the front end stops this empty username and password being handed in
        // still check here just in case
        if (username === '' || password === '') {
            // Will be changed to a user notification in the future
            console.error('Username and password must not be empty');
            return;
        }
        try {
            const user_login = await login({username, password});
            setUser(user_login);
            setUsername('');
            setPassword('');
            console.log('Logged in:', user);
        } catch (exception) {
            console.error(exception);
        }
    }

    return (
        <form onSubmit={handleLogin}>
            <Input
                isRequired
                type="username"
                label="Username"
                variant="bordered"
                className="max-w-xs"
                onChange={({ target }) => setUsername(target.value)}
            />
            <Input
                isRequired
                label="Password"
                variant="bordered"
                endContent={
                    <button className="focus:outline-none" type="button" onClick={toggleVisibility}>
                        {isVisible ? (
                            <EyeSlashFilledIcon className="text-2xl text-default-400 pointer-events-none" />
                        ) : (
                            <EyeFilledIcon className="text-2xl text-default-400 pointer-events-none" />
                        )}
                    </button>
                }
                type={isVisible ? "text" : "password"}
                className="max-w-xs"
                onChange={({ target }) => setPassword(target.value)}
            />
            <Button color="primary" type="submit" data-testid="login-button">Login</Button>
        </form>
    )
}
