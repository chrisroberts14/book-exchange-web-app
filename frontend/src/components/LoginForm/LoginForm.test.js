import {LoginForm} from "./LoginForm.jsx";
import {fireEvent, render, screen, waitFor} from '@testing-library/react';
import { login } from '../../api/api_calls';
import { userEvent } from '@testing-library/user-event';
import {act} from "react";

jest.mock('../../api/api_calls');


describe('LoginForm', () => {
    let setUser;

    beforeEach(() => {
        setUser = jest.fn();
        console.log = jest.fn();
        console.error = jest.fn();
    });

    it('renders correctly', () => {
        const { container } = render(<LoginForm />);
        expect(container).toMatchSnapshot();
    });

    it('Login submits correctly', async () => {
        login.mockResolvedValue({ username: 'test_user' });

        render(
            <LoginForm
                user={"sample_user"}
                setUser={setUser}
            />
        );
        // Simulate user input
        await userEvent.type(screen.getByLabelText('Username'), 'test_user');
        await userEvent.type(screen.getByLabelText('Password'), 'password');

        fireEvent.click(screen.getByTestId('login-button'));

        await waitFor(() => {
            expect(login).toHaveBeenCalledWith({ username: 'test_user', password: 'password' });
        })
    });

    it('Login fails correctly', async () => {
       // For now this is a console.error but in the future this will be a notification
        login.mockImplementation(() => {
           throw new Error();
        });

        render(
            <LoginForm
                user={"sample_user"}
                setUser={setUser}
            />
        );
        // Simulate user input
        await userEvent.type(screen.getByLabelText('Username'), 'test_user');
        await userEvent.type(screen.getByLabelText('Password'), 'password');

        fireEvent.click(screen.getByTestId('login-button'));

        expect(String(console.error.mock.calls[0])).toStrictEqual("Error");
    });

    it('Login fails correctly with empty password and or username', async () => {
        render(
            <LoginForm
                user={"sample_user"}
                setUser={setUser}
            />
        );

        fireEvent.click(screen.getByTestId('login-button'));

        expect(String(console.error.mock.calls[0])).toStrictEqual("Username and password must not be empty");
    });

});
