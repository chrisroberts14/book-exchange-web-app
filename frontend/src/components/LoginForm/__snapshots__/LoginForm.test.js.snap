// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders correctly 1`] = `
<div>
  <form>
    <div
      class="group flex flex-col w-full max-w-xs"
      data-has-elements="true"
      data-has-label="true"
      data-has-value="true"
      data-required="true"
      data-slot="base"
    >
      <div
        class="relative w-full inline-flex tap-highlight-transparent shadow-sm px-3 border-medium border-default-200 data-[hover=true]:border-default-400 group-data-[focus=true]:border-default-foreground min-h-10 rounded-medium flex-col items-start justify-center gap-0 transition-background !duration-150 transition-colors motion-reduce:transition-none h-14 py-2"
        data-slot="input-wrapper"
        style="cursor: text;"
      >
        <label
          class="absolute z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 cursor-text after:content-['*'] after:text-danger after:ml-0.5 rtl:after:ml-[unset] rtl:after:mr-0.5 will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 text-small group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.small)/2_-_6px_-_theme(borderWidth.medium))] pe-2 max-w-full text-ellipsis overflow-hidden"
          data-slot="label"
          for="react-aria-:r0:"
          id="react-aria-:r1:"
        >
          Username
        </label>
        <div
          class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end"
          data-slot="inner-wrapper"
        >
          <input
            aria-label="Username"
            aria-labelledby="react-aria-:r0: react-aria-:r1:"
            aria-required="true"
            class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 text-small"
            data-slot="input"
            id="react-aria-:r0:"
            type="username"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="group flex flex-col w-full max-w-xs"
      data-has-elements="true"
      data-has-label="true"
      data-has-value="true"
      data-required="true"
      data-slot="base"
    >
      <div
        class="relative w-full inline-flex tap-highlight-transparent shadow-sm px-3 border-medium border-default-200 data-[hover=true]:border-default-400 group-data-[focus=true]:border-default-foreground min-h-10 rounded-medium flex-col items-start justify-center gap-0 transition-background !duration-150 transition-colors motion-reduce:transition-none h-14 py-2"
        data-slot="input-wrapper"
        style="cursor: text;"
      >
        <label
          class="absolute z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 cursor-text after:content-['*'] after:text-danger after:ml-0.5 rtl:after:ml-[unset] rtl:after:mr-0.5 will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 text-small group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.small)/2_-_6px_-_theme(borderWidth.medium))] pe-2 max-w-full text-ellipsis overflow-hidden"
          data-slot="label"
          for="react-aria-:r5:"
          id="react-aria-:r6:"
        >
          Password
        </label>
        <div
          class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end"
          data-slot="inner-wrapper"
        >
          <input
            aria-label="Password"
            aria-labelledby="react-aria-:r5: react-aria-:r6:"
            aria-required="true"
            class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 text-small"
            data-has-end-content="true"
            data-slot="input"
            id="react-aria-:r5:"
            type="password"
            value=""
          />
          <button
            class="focus:outline-none"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="text-2xl text-default-400 pointer-events-none"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M21.25 9.14969C18.94 5.51969 15.56 3.42969 12 3.42969C10.22 3.42969 8.49 3.94969 6.91 4.91969C5.33 5.89969 3.91 7.32969 2.75 9.14969C1.75 10.7197 1.75 13.2697 2.75 14.8397C5.06 18.4797 8.44 20.5597 12 20.5597C13.78 20.5597 15.51 20.0397 17.09 19.0697C18.67 18.0897 20.09 16.6597 21.25 14.8397C22.25 13.2797 22.25 10.7197 21.25 9.14969ZM12 16.0397C9.76 16.0397 7.96 14.2297 7.96 11.9997C7.96 9.76969 9.76 7.95969 12 7.95969C14.24 7.95969 16.04 9.76969 16.04 11.9997C16.04 14.2297 14.24 16.0397 12 16.0397Z"
                fill="currentColor"
              />
              <path
                d="M11.9984 9.14062C10.4284 9.14062 9.14844 10.4206 9.14844 12.0006C9.14844 13.5706 10.4284 14.8506 11.9984 14.8506C13.5684 14.8506 14.8584 13.5706 14.8584 12.0006C14.8584 10.4306 13.5684 9.14062 11.9984 9.14062Z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <button
      class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary text-primary-foreground data-[hover=true]:opacity-hover"
      data-testid="login-button"
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
